Extract all access of APIv3 calls into APIv3Service

oliverguenther opened the pull request:
This refactoring will make it easier for developers to discover functionality and connected services with the APIv3.

What this does:

- Create an entry point `ApiV3Service` that provides path helpers AND accessors to functioanlity in the API. This pattern was first introduced with the [BCFApiService](https://github.com/opf/openproject/blob/dev/frontend/src/app/modules/bim/bcf/api/bcf-api.service.ts#L42) that too allows for navigation of resources AND accessing or modifying them directly.

- Remove all `*DmService` implementations
- Remove the `apiV3Path` implementation
- Remove all directly accessible `StateCacheService`s with the exception of the schemaCache (since used on many different endpoints) and the relations cache (basically a service to aggregate relations for a single work package)
- Avoids having to inject a variety of services to get access or download a HAL resource
- Caches are contained within their respective resource without injecting another service. There is still the `States` service to access these states directly (however not recommended)
- It is no longer necessary to manually update the cache after executing a `POST` / `PATCH` , ... request since the apiv3 Service will automatically update the cache with the change.

Developers can now access a work package by doing the following:

```typescript
this
  .apiV3Service
  .work_packages.id("1234")
  .get() // Observable<WorkPackageResource>
```

and will automatically receive an observable to a work package resource that may or may not have come from a cache.
What was `wpCacheService.require` is now

```typescript
this
  .apiV3Service
  .work_packages.id("1234")
  .requireAndStream() // Observable<WorkPackageResource>
```

which loads the initial value and then stream all cached values.

oliverguenther committed:
Move apiv3 path helper into apiv3 service

oliverguenther committed:
Refactor application to use apiv3 service

oliverguenther committed:
Allow caching responses within apiv3 endpoints through state cache

openprojectci commented on frontend/src/app/modules/boards/board/inline-add/board-inline-add-autocompleter.component.ts:
```diff
@@ -51,6 +51,7 @@ import {SchemaCacheService} from "core-components/schemas/schema-cache.service";
 import {WorkPackageNotificationService} from "core-app/modules/work_packages/notifications/work-package-notification.service";
 import {UrlParamsHelperService} from "core-components/wp-query/url-params-helper";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/boards/board/inline-add/board-inline-add-autocompleter.component.ts#L55"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/common/autocomplete/user-autocompleter.component.ts:
```diff
@@ -37,6 +37,7 @@ import {DebouncedRequestSwitchmap, errorNotificationHandler} from "core-app/help
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
 export const usersAutocompleterSelector = 'user-autocompleter';
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/common/autocomplete/user-autocompleter.component.ts#L43"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/common/autocomplete/version-autocompleter.component.ts:
```diff
@@ -34,6 +34,7 @@ import {VersionResource} from "core-app/modules/hal/resources/version-resource";
 import {I18nService} from "core-app/modules/common/i18n/i18n.service";
 import {HalResourceNotificationService} from "core-app/modules/hal/services/hal-resource-notification.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/cfe89dd706c72154281f341454501767d598e770/frontend/src/app/modules/common/autocomplete/version-autocompleter.component.ts#L38"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/job-status/job-status-modal/job-status.modal.ts:
```diff
@@ -13,6 +13,7 @@ import {
 import {NotificationsService} from "core-app/modules/common/notifications/notifications.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/job-status/job-status-modal/job-status.modal.ts#L18"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther committed:
Fix reference to root service

oliverguenther committed:
Continue refactoring of DM services

oliverguenther committed:
Refactor version DM service

openprojectci commented on frontend/src/app/components/modals/editor/macro-wp-button-modal/wp-button-macro.modal.ts:
```diff
@@ -40,7 +40,8 @@ import {OpModalLocalsMap} from "core-components/op-modals/op-modal.types";
-import {WorkPackageDmService} from "core-app/modules/hal/dm-services/work-package-dm.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
+import {FormResource} from "core-app/modules/hal/resources/form-resource";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/modals/editor/macro-wp-button-modal/wp-button-macro.modal.ts#L45"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/common/autocomplete/version-autocompleter.component.ts:
```diff
@@ -27,13 +27,13 @@
 import {I18nService} from "core-app/modules/common/i18n/i18n.service";
 import {HalResourceNotificationService} from "core-app/modules/hal/services/hal-resource-notification.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/common/autocomplete/version-autocompleter.component.ts#L37"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther committed:
Refactor order service into v3 paths

oliverguenther committed:
Replace time entry dm and cache service

openprojectci commented on frontend/src/app/modules/fields/edit/field-types/te-work-package-edit-field.component.ts:
```diff
@@ -28,22 +28,22 @@
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
 const RECENT_TIME_ENTRIES_MAGIC_NUMBER = 30;
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/fields/edit/field-types/te-work-package-edit-field.component.ts#L41"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented:
<!--
  0 Errors
  4 Warnings: <a href="https://github.com/op...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-relations/wp-relations-create/wp-relations-create.component.ts#L8">frontend/src/app/components/wp-relations/wp-relations-create/wp-relations-create.component.ts#L8</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/boards/board/add-list-modal/add-list-modal.component.ts#L42">frontend/src/app/modules/boards/board/add-list-modal/add-list-modal.component.ts#L42</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts#L40">frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts#L40</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/grids/widgets/time-entries/project/time-entries-project.component.ts#L9">frontend/src/app/modules/grids/widgets/time-entries/project/time-entries-project.component.ts#L9</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

oliverguenther committed:
Replace QueryFormDmService with nested form resource

oliverguenther committed:
Replace QueryDmService with api paths

oliverguenther committed:
Replace ProjectDM and ProjectCacheService

oliverguenther committed:
Move PayloadDmService into static helper class

oliverguenther committed:
Replace NewsDmService with api paths

oliverguenther committed:
Replace MembershipDmService with api paths

oliverguenther committed:
Replace HelpTextDmService with api paths

openprojectci commented on frontend/src/app/components/wp-table/configuration-modal/wp-table-configuration.modal.ts:
```diff
@@ -33,6 +32,7 @@ import {I18nService} from "core-app/modules/common/i18n/i18n.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
 export const WpTableConfigurationModalPrependToken = new InjectionToken<ComponentType<any>>('WpTableConfigurationModalPrependComponent');
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-table/configuration-modal/wp-table-configuration.modal.ts#L38"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-table/embedded/wp-embedded-table.component.ts:
```diff
@@ -4,15 +4,14 @@ import {WorkPackageViewPaginationService} from 'core-app/modules/work_packages/r
 import {InjectField} from "core-app/helpers/angular/inject-field.decorator";
 import {KeepTabService} from "core-components/wp-single-view-tabs/keep-tab/keep-tab.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-table/embedded/wp-embedded-table.component.ts#L15"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/boards/board/board-filter/board-filter.component.ts:
```diff
@@ -15,6 +14,7 @@ import {debounceTime, skip, take} from "rxjs/internal/operators";
 import {Observable} from "rxjs";
 import {BoardFiltersService} from "core-app/modules/boards/board/board-filter/board-filters.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/boards/board/board-filter/board-filter.component.ts#L18"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther committed:
Replace ConfigurationDmService with apiv3 path

oliverguenther committed:
Replace GridDmService with paths

oliverguenther committed:
Remove DmListParameters

This is now Apiv3ListParameters type

oliverguenther committed:
Replace virtual BoardDmService with a virtual path

Boards are not actually their own API resource, but a shallow wrapper
over a grid resource.

To reuse the same API service pattern, we can allow other resources
other than HalResource and make the boards path just reference the grids
paths.

This allows to nicely use the `apiV3Service.boards.id(..).cache`
patterns.

oliverguenther committed:
Refactor cacheState into single method

oliverguenther committed:
Make HalEditingService a v3 cache without loading ability

oliverguenther committed:
Replace usages of WorkPackageCacheService with api-integrated cache

oliverguenther committed:
Fix accessing grid subpath

oliverguenther committed:
Correctly use schema ensureLoaded before inserting into caches

openprojectci commented on frontend/src/app/components/filters/quick-filter-by-text-input/quick-filter-by-text-input.component.ts:
```diff
@@ -29,13 +29,13 @@
 import {QueryFilterResource} from "core-app/modules/hal/resources/query-filter-resource";
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/filters/quick-filter-by-text-input/quick-filter-by-text-input.component.ts#L39"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/work-packages/wp-relations-count/wp-relations-count.component.ts:
```diff
@@ -1,8 +1,8 @@
 import {combineLatest} from 'rxjs';
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/work-packages/wp-relations-count/wp-relations-count.component.ts#L6"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/work-packages/wp-relations-count/wp-watchers-count.component.ts:
```diff
@@ -1,8 +1,8 @@
 import { WorkPackageWatchersService } from 'core-app/components/wp-single-view-tabs/watchers-tab/wp-watchers.service';
 import { HalResource } from 'core-app/modules/hal/resources/hal-resource';
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e71869ec897476f98b23dceb4c3c60722bb3632a/frontend/src/app/components/work-packages/wp-relations-count/wp-watchers-count.component.ts#L6"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/work-packages/wp-subject/wp-subject.component.ts:
```diff
@@ -29,9 +29,9 @@
 import {randomString} from "core-app/helpers/random-string";
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/work-packages/wp-subject/wp-subject.component.ts#L35"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/work-packages/wp-watcher-button/wp-watcher-button.component.ts:
```diff
@@ -27,11 +27,11 @@
 import {WorkPackageWatchersService} from 'core-components/wp-single-view-tabs/watchers-tab/wp-watchers.service';
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e71869ec897476f98b23dceb4c3c60722bb3632a/frontend/src/app/components/work-packages/wp-watcher-button/wp-watcher-button.component.ts#L35"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-buttons/wp-status-button/wp-status-button.component.ts:
```diff
@@ -33,12 +33,9 @@ import {ChangeDetectorRef, Component, Input, OnInit} from '@angular/core';
 import {SchemaCacheService} from "core-components/schemas/schema-cache.service";
-import {ISchemaProxy, SchemaProxy} from "core-app/modules/hal/schemas/schema-proxy";
+import {ISchemaProxy} from "core-app/modules/hal/schemas/schema-proxy";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-buttons/wp-status-button/wp-status-button.component.ts#L39"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-custom-actions/wp-custom-actions/wp-custom-action.component.ts:
```diff
@@ -38,6 +37,7 @@ import {HalResourceEditingService} from "core-app/modules/fields/edit/services/h
 import {HalEventsService} from "core-app/modules/hal/services/hal-events.service";
 import {WorkPackageNotificationService} from "core-app/modules/work_packages/notifications/work-package-notification.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-custom-actions/wp-custom-actions/wp-custom-action.component.ts#L41"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-relations/embedded/children/wp-children-query.component.ts:
```diff
@@ -37,17 +37,17 @@ import {OpTableActionFactory} from 'core-components/wp-table/table-actions/table
 import {GroupDescriptor} from "core-components/work-packages/wp-single-view/wp-single-view.component";
 import {HalEventsService} from "core-app/modules/hal/services/hal-events.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-relations/embedded/children/wp-children-query.component.ts#L45"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-relations/embedded/inline/add-existing/wp-relation-inline-add-existing.component.ts:
```diff
@@ -40,6 +39,7 @@ import {UrlParamsHelperService} from "core-components/wp-query/url-params-helper
 import {HalEventsService} from "core-app/modules/hal/services/hal-events.service";
 import {WorkPackageNotificationService} from "core-app/modules/work_packages/notifications/work-package-notification.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-relations/embedded/inline/add-existing/wp-relation-inline-add-existing.component.ts#L43"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-relations/wp-relation-row/wp-relation-row.component.ts:
```diff
@@ -8,6 +7,7 @@ import {I18nService} from "core-app/modules/common/i18n/i18n.service";
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-relations/wp-relation-row/wp-relation-row.component.ts#L12"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-relations/wp-relations-hierarchy/wp-relations-hierarchy.directive.ts:
```diff
@@ -32,8 +32,8 @@ import {WorkPackageResource} from 'core-app/modules/hal/resources/work-package-r
-import {WorkPackageCacheService} from '../../work-packages/work-package-cache.service';
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-relations/wp-relations-hierarchy/wp-relations-hierarchy.directive.ts#L37"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-single-view-tabs/overview-tab/overview-tab.component.ts:
```diff
@@ -28,10 +28,10 @@
 import {I18nService} from 'core-app/modules/common/i18n/i18n.service';
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-single-view-tabs/overview-tab/overview-tab.component.ts#L35"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-single-view-tabs/relations-tab/relations-tab.component.ts:
```diff
@@ -27,11 +27,11 @@
 import {WorkPackageResource} from 'core-app/modules/hal/resources/work-package-resource';
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/components/wp-single-view-tabs/relations-tab/relations-tab.component.ts#L35"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/boards/board/board-partitioned-page/board-list-container.component.ts:
```diff
@@ -21,6 +20,7 @@ import {AddListModalComponent} from "core-app/modules/boards/board/add-list-moda
 import {BoardListCrossSelectionService} from "core-app/modules/boards/board/board-list/board-list-cross-selection.service";
 import {filter} from "rxjs/operators";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/boards/board/board-partitioned-page/board-list-container.component.ts#L24"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/boards/board/configuration-modal/board-configuration.modal.ts:
```diff
@@ -23,7 +23,6 @@ import {OpModalLocalsToken} from "core-components/op-modals/op-modal.service";
 import {Board} from "core-app/modules/boards/board/board";
-import {BoardCacheService} from "core-app/modules/boards/board/board-cache.service";
 
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/boards/board/configuration-modal/board-configuration.modal.ts#L27"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/boards/boards-sidebar/boards-menu.component.ts:
```diff
@@ -2,12 +2,12 @@ import {Component, OnInit} from "@angular/core";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
 export const boardsMenuSelector = 'boards-menu';
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/boards/boards-sidebar/boards-menu.component.ts#L13"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/boards/index-page/boards-index-page.component.ts:
```diff
@@ -15,6 +14,7 @@ import {DomSanitizer} from "@angular/platform-browser";
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
 import {componentDestroyed} from "@w11k/ngx-componentdestroyed";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/e6d2da33c82cc74f79acd1ed35cd9de99459dca5/frontend/src/app/modules/boards/index-page/boards-index-page.component.ts#L18"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther committed:
Fix query not having an ID when updating

oliverguenther committed:
Fix npm tests after removal of wp cache

oliverguenther committed:
Readd missing path helper for APIv3 used from ckeditor

oliverguenther committed:
Correctly save query in widget

oliverguenther committed:
Fix watcher tab

oliverguenther committed:
Fix search path

oliverguenther committed:
Use a promise for the updateImmediately link of time entry

oliverguenther committed:
Make apiv3 independent cache services depend on the base cache service

We had two cache services now, one base service for the apiv3, and the
previous one. Both shared a lot of code, so we can unify them

oliverguenther committed:
Don't use multiState getter in schemaCache

oliverguenther committed:
Correct loading of schema

oliverguenther committed:
Correctly fill state with relations

oliverguenther committed:
Don't render comment until workPackage ready

oliverguenther committed:
Merge the initial loading observable to get error handling

If we only return state.values$(),
we never get the initial http observable and thus no error handling

oliverguenther committed:
Ensure we reload the default query from the current project if any

oliverguenther committed:
Use get() instead of stream

oliverguenther committed:
Load types correctly

oliverguenther committed:
Fix paths in enterprise token urls

oliverguenther committed:
Ensure take(1) in requireAndStream toPromise usage

oliverguenther committed:
Avoid duplicate StateCacheService definition

oliverguenther committed:
Only use project based query when there is no query.id yet

There is no endpoint for /api/v3/projects/:projectIdentifier/queries/:id

oliverguenther committed:
Add isAvailable to filter out filters incorrectly added

The search component always adds a subproject filter when in project scope.
This filter however is only valid/existent in schemas if there is a subproject.

This ensures we only try to access filters that are available

oliverguenther committed:
Refactor embedded table to reuse the loadQueryFromExisting method

We had a few duplicate methods on how to look for a query with params
basing on an existing (saved or unsaved) query.

oliverguenther committed:
Ensure filter schemas are inserted and checked before use

oliverguenther committed:
Better matcher for search spec

oliverguenther committed:
Ensure correct initialization of query states

oliverguenther committed:
Load attribute help text correctly

huornlmj commented:
According to LGTM.com, this merge introduced 77 new issues, all of the 'unused import' category.
![image](https://user-images.githubusercontent.com/28444739/116998256-d6ff5000-acd5-11eb-983b-5f1bd438f037.png)
